### PR TITLE
Complex Synthetic Feature Generation

### PR DESCRIPTION
**Pull request by 98MM**
Added a suite of functions related to synthetic feature generation and corresponding tests.

Located in outrank/algorithms/synthetic_data_generators

_https://jira.outbrain.com/browse/mm_generators_

### CODE REVIEW

**Comment by 98MM** [comment]
HTML docs are yet to be written.

**Review by SkBlaz — CHANGES_REQUESTED** [review]
`inspectionProfiles` -> we don't need that. Please remove all files that are not relevant to the actual PR's content. Also, ccgen_tests.py and cc_generator.py seem same?

**Comment by 98MM** [comment]
> `inspectionProfiles` -> we don't need that. Please remove all files that are not relevant to the actual PR's content. Also, ccgen_tests.py and cc_generator.py seem same?

Removed .idea, cc_generator and cc_generator_tests are differet files, i've checked.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 1]
```diff
@@ -0,0 +1,702 @@
+import numpy as np
+from scipy.linalg import qr
+from scipy.stats import norm
+from sklearn.cluster import KMeans
+from sklearn.utils import resample
+from typing import List, Union, Optional, Tuple
+class CategoricalClassification:
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:7]: Did you install `pre-commit` hook? Please run it, spacing looks off

Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py:15]: pre-commit installed, ran

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+            for i in range(n_features):
+                x = self._generate_feature(cardinality, n_samples, ensure_rep=ensure_rep)
+                X[i] = x
+
+        else:
+
+            ix = 0
+            for data in structure:
+
+                if not isinstance(data[0], (list, np.ndarray)):
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py:87]: used pre-commit, fixed issue

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+                'structure': structure,
+                'ensure_rep': ensure_rep,
+                'seed': seed
+            }
+        })
+
+        np.random.seed(seed)
+        X = np.empty([n_features, n_samples])
+
+        if structure == None:
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: replaced with structure is None

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+                    feature_cardinality = data[1]
+                    for feature_ix in feature_ixs:
+                        if ix < feature_ix:
+                            for i in range(ix, feature_ix):
+                                x = self._generate_feature(cardinality, n_samples, ensure_rep=ensure_rep)
+                                X[ix] = x
+                                ix += 1
+
+                        if not isinstance(feature_cardinality, (list, np.ndarray)):
+                            x = self._generate_feature(feature_cardinality, n_samples, ensure_rep=ensure_rep)
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Extracted to _feature_buildetr method

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 5]
```diff
@@ -0,0 +1,702 @@
...
+                            x = self._generate_feature(cardinality, n_samples, ensure_rep=ensure_rep)
+                            X[ix] = x
+                            ix += 1
+
+                    if not isinstance(feature_cardinality, (list, np.ndarray)):
+                        x = self._generate_feature(feature_cardinality, n_samples, ensure_rep=ensure_rep)
+                    else:
+                        if isinstance(feature_cardinality[0], (list, np.ndarray)):
+                            value_domain = feature_cardinality[0]
+                            value_frequencies = feature_cardinality[1]
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Replaced all occurences with single line unpacking

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 6]
```diff
@@ -0,0 +1,702 @@
...
+        return X.T
+
+    def _generate_feature(self,
+                          v: Union[int, List[int], np.ndarray],
+                          size: int,
+                          ensure_rep: bool = False,
+                          p: Optional[Union[List[float], np.ndarray]] = None) -> np.ndarray:
+        """
+        Generates feature vector of length size. Default probability density distribution is approx. normal, centred around randomly picked value.
+        :param v: either int for cardinality, or list of values
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Defined interface as suggested, updated function.

**Comment by 98MM** [comment]
> I'll continue review later. Some comments first.

Sorry for not replying earlier, had implemented the necessary changes with the changelog in my commit description, should have replied to the comments as well. Thank you for all the feedback.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 7]
```diff
@@ -45,113 +61,214 @@ def generate_data(self,
...
                     X[i] = x
 
         return X.T
 
-    def _generate_feature(self,
-                          v: Union[int, List[int], np.ndarray],
-                          size: int,
-                          ensure_rep: bool = False,
-                          p: Optional[Union[List[float], np.ndarray]] = None) -> np.ndarray:
+    def _feature_builder(
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Renamed _feature_builder to _configure_generate_feature (as it what it does based on the feature_attributes parameter). Cleaned up the docstring.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 8]
```diff
@@ -45,113 +61,214 @@ def generate_data(self,
...
+        self,
+        feature_attributes: int | list | np.ndarray,
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Replced ``np.ndarray`` with ``ArrayLike`` from the NumPy Typing package.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 9]
```diff
@@ -45,113 +61,214 @@ def generate_data(self,
...
+        ensure_rep: bool = False,
+        random_values: bool | None = False,
+        low: int | None = 0,
+        high: int | None = 1000,
+    ) -> np.ndarray:
+
         """
-        Generates feature vector of length size. Default probability density distribution is approx. normal, centred around randomly picked value.
-        :param v: either int for cardinality, or list of values
+        Helper function to avoid duplicate code, builds feature
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Rewrote docstring to better reflect function functionality.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 10]
```diff
@@ -493,12 +622,14 @@ def _cluster_data(self,
...
-                       y: Union[List[int], np.ndarray],
-                       p: float = 0.2,
-                       type: str = "categorical",
-                       missing_val: Union[str, int, float] = float('-inf')) -> np.ndarray:
+    def generate_noise(
+        self,
+        X: np.ndarray,
+        y: list[int] | np.ndarray,
+        p: float = 0.2,
+        type: str = 'categorical',
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Replaced all ``str`` typing hints with ``Literal``.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 11]
```diff
@@ -0,0 +1,845 @@
...
+
+        np.random.shuffle(sampled_values)
+        return sampled_values
+
+    def generate_combinations(
+        self,
+        X: ArrayLike,
+        feature_indices: list[int] | ArrayLike,
+        combination_function: Optional = None,
+        combination_type: Literal = 'linear',
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Fixed, thank you

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 12]
```diff
@@ -0,0 +1,845 @@
...
+                    overflow_indices = np.delete(overflow_indices, sample_indices_slice, axis=0)
+
+        return np.array(cluster_labels)
+
+    def generate_noise(
+        self,
+        X: ArrayLike,
+        y: list[int] | ArrayLike,
+        p: float = 0.2,
+        type: Literal = 'categorical',
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Fixed

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 13]
```diff
@@ -0,0 +1,845 @@
...
+
+                    feature[ix] = val
+
+            rev_ind = inds.argsort()
+            X_noise = Xs_T.T
+            X_noise = X_noise[rev_ind]
+
+            return X_noise
+
+        elif type == 'missing':
```
[the remark this answers is not in the document]
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py:702]: Added ValueError, f'{type} not supported.'

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 14]
```diff
@@ -0,0 +1,849 @@
...
+            X_noise = np.copy(X)
+            Xn_T = X_noise.T
+            n = Xn_T.shape[1]
+            n_missing = int(n * p)
+            #print("n to delete:", n_missing)
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:707]: Redundant comment

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 15]
```diff
@@ -0,0 +1,849 @@
...
+
+            for feature in Xn_T:
+                ixs = np.random.choice(n, n_missing, replace=False)
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:710]: Make sure seeds are set for numpy

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 16]
```diff
@@ -0,0 +1,849 @@
...
+
+                for ix in ixs:
+                    feature[ix] = missing_val
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:713]: You in general do a lot of for loop-based indexing -- this is really not the optimal way, this can all be vectorized. Not in this PR thought, let's go over separately sometime

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 17]
```diff
@@ -0,0 +1,849 @@
...
+        print(f"Number of generated samples: {self.dataset_info['general']['n_samples']}")
+        if 'downsampling' in self.dataset_info:
+            print(
+                f"Dataset downsampled from shape {self.dataset_info['downsampling']['original_shape']}, to shape {self.dataset_info['downsampling']['downsampled_shape']}",
+            )
+        if 'n_class' in self.dataset_info['labels']:
+            print(f"Number of classes: {self.dataset_info['labels']['n_class']}")
+            print(f"Class relation: {self.dataset_info['labels']['class_relation']}")
+
+        print('-------------------------------------')
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:817]: Let's leave this for now, but outrank has proper logging which should be used when time comes, please put as todo

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 18]
```diff
@@ -0,0 +1,849 @@
...
+            adjustments = []
+            overflow_samples = []
+            overflow_indices = []
+            for i in range(n):
+                cluster_size = np.sum(cluster_labels == i)
+
+                adjustment = samples_per_cluster[i] - cluster_size
+                adjustments.append(adjustment)
+
+                if adjustment < 0:  # Cluter is too large
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:575]: Comments always above lines, not after pls

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 19]
```diff
@@ -0,0 +1,849 @@
...
+        """
+        Cluster data using kmeans
+        :param X: dataset
+        :param n: number of clusters
+        :param p: class distribution
+        :param balance: balance the clusters according to p
+        :return: array of labels, corresponding to dataset X
+        """
+
+        kmeans = KMeans(n_clusters=n)
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:543]: Seed for kmeans set?

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 20]
```diff
@@ -0,0 +1,849 @@
...
+                    X[ix] = x
+                    ix += 1
+
+                else:
+                    # Data in structure is a tuple of (list of feature indexes, feature attributes)
+                    feature_ixs, feature_attributes = data
+
+                    for feature_ix in feature_ixs:
+                        # Filling out the dataset up to feature_ix
+                        if ix < feature_ix:
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:124]: Also a general comment - it's not math source, you can name variables with fuller/more idiomatic names, not this PR though

File: outrank/algorithms/synthetic_data_generators/cc_generator.py [hunk 21]
```diff
@@ -0,0 +1,849 @@
...
+
+        np.random.shuffle(sampled_values)
+        return sampled_values
+
+    def generate_combinations(
+        self,
+        X: ArrayLike,
+        feature_indices: list[int] | ArrayLike,
+        combination_function: Optional = None,
+        combination_type: Literal['linear', 'nonlinear'] = 'linear',
```
Comment (SkBlaz) [inline outrank/algorithms/synthetic_data_generators/cc_generator.py:271]: Looks like `str` to me - @miha-jenko why Literal?

**Review by SkBlaz — CHANGES_REQUESTED** [review]
@98MM few smaller changes, and we should be good to go.  I'll add unit tests to CI too soon

File: outrank/algorithms/synthetic_data_generators/cc_generator.py (hunk 14 above)
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py:715]: Remoevd redundant comments

File: outrank/algorithms/synthetic_data_generators/cc_generator.py (hunk 15 above)
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py:718]: Seeds are now set when class is instantiated  with ```cc = CategoricalClassification()```, default is 42,  pass ```seed``` parameter to change.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py (hunk 17 above)
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Removed summarize function, added TODO

File: outrank/algorithms/synthetic_data_generators/cc_generator.py (hunk 18 above)
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Moved all comments above code block they're referencing.

File: outrank/algorithms/synthetic_data_generators/cc_generator.py (hunk 19 above)
Comment (98MM) [reply outrank/algorithms/synthetic_data_generators/cc_generator.py]: Added ```random_state``` parameter for setting clustering seed.

**Review by SkBlaz — APPROVED** [review]
